pkg/util/worker.go

```diff
@@ -54,46 +54,61 @@ func (w *worker[T]) Do(ctx context.Context, r func(WorkItemResult[T]), wrk WorkI
 	}
 }
 
+type GroupedItem[T any] struct {
+	Group int
+	Item  WorkItem[T]
+}
+
 type WorkerGroup[T any] struct {
 	maxWorkers    int
 	activeWorkers int
 	workers       chan *worker[T]
-	queue         chan WorkItem[T]
-	queuedItems   atomic.Int64
-	queueClosed   atomic.Bool
-	drain         chan struct{}
-	svcCtx        context.Context
-	svcCancel     context.CancelFunc
-	resultData    []WorkItemResult[T]
-	resultNotify  chan struct{}
-	resultLen     atomic.Int64
-	mu            sync.RWMutex
-	once          sync.Once
+
+	queue         *Queue[GroupedItem[T]]
+	input         chan GroupedItem[T]
+	chInputNotify chan struct{}
+	// TODO: make result data a cache instead of a map
+	mu           sync.Mutex
+	resultData   map[int][]WorkItemResult[T]
+	resultNotify map[int]chan struct{}
+
+	// channels used to stop processing
+	chStopInputs     chan struct{}
+	chStopProcessing chan struct{}
+	queueClosed      atomic.Bool
+
+	// service state management
+	svcCtx    context.Context
+	svcCancel context.CancelFunc
+	once      sync.Once
 }
 
 func NewWorkerGroup[T any](workers int, queue int) *WorkerGroup[T] {
 	svcCtx, svcCancel := context.WithCancel(context.Background())
 	wg := &WorkerGroup[T]{
-		maxWorkers:   workers,
-		workers:      make(chan *worker[T], workers),
-		queue:        make(chan WorkItem[T], queue),
-		drain:        make(chan struct{}, 1),
-		resultData:   make([]WorkItemResult[T], 0),
-		resultNotify: make(chan struct{}, 1),
-		svcCtx:       svcCtx,
-		svcCancel:    svcCancel,
+		maxWorkers:       workers,
+		workers:          make(chan *worker[T], workers),
+		queue:            &Queue[GroupedItem[T]]{},
+		input:            make(chan GroupedItem[T], 1),
+		chInputNotify:    make(chan struct{}, 1),
+		resultData:       map[int][]WorkItemResult[T]{},
+		resultNotify:     map[int]chan struct{}{},
+		chStopInputs:     make(chan struct{}),
+		chStopProcessing: make(chan struct{}),
+		svcCtx:           svcCtx,
+		svcCancel:        svcCancel,
 	}
 
 	go wg.run()
 
-	runtime.SetFinalizer(wg, func(g *WorkerGroup[T]) { g.Stop() })
+	// runtime.SetFinalizer(wg, func(g *WorkerGroup[T]) { g.Stop() })
```

Comment:
I guess we can keep this line in until we decide what's happening with finalizers/the plugin lifecycle?